File: src/slides_thief/detection/hough_lines.py
"""Orientation-guided probabilistic Hough line and free-direction quad generation."""

from __future__ import annotations

from dataclasses import dataclass
import math

import numpy as np

from .config import DETECTION_CONFIG
from .gradient import GradientMap
from .geometry import geometry_is_valid, line_intersection, normalized_corner_distance, polygon_area
from .numeric import average

_HOUGH_CONFIG = DETECTION_CONFIG["houghLines"]

ANGLE_STEP = math.pi / int(_HOUGH_CONFIG["angleBins"])
RHO_STEP = float(_HOUGH_CONFIG["rhoStep"])


@dataclass
class Segment:
    line: np.ndarray
    angle: float
    normal_angle: float
    rho: float
    length: float
    support: float


def hough_quad_candidates(gradient: GradientMap) -> list[dict]:
    height, width = gradient.magnitude.shape
    points = _edge_points(gradient)
    if len(points) < int(_HOUGH_CONFIG["minimumEdgePoints"]):
        return []
    peaks = _vote(points, width, height)
    segments = _extract_segments(peaks, points, width, height)
    families = _cluster_directions(segments)
    if families is None:
        return []
    first_family, second_family = families
    family_angle = _orientation_difference(first_family[0], second_family[0])
    if (
        family_angle < math.radians(float(_HOUGH_CONFIG["minimumFamilyAngleDegrees"]))
        or family_angle > math.radians(float(_HOUGH_CONFIG["maximumFamilyAngleDegrees"]))
    ):
        return []

    first_pairs = _line_pairs(first_family[1], width, height)
    second_pairs = _line_pairs(second_family[1], width, height)
    candidates = []
    for first_pair in first_pairs:
        for second_pair in second_pairs:
            corners = [
                line_intersection(first_pair[0].line, second_pair[0].line),
                line_intersection(first_pair[0].line, second_pair[1].line),
                line_intersection(first_pair[1].line, second_pair[1].line),
                line_intersection(first_pair[1].line, second_pair[0].line),
            ]
            quad = _order_quad(np.asarray(corners, dtype=np.float64))
            if not geometry_is_valid(quad, width, height):
                continue
            area = polygon_area(quad) / (width * height)
            supports = [segment.support for segment in (*first_pair, *second_pair)]
            candidates.append(
                {
                    "quad": quad,
                    "method": "hough-lines",
                    "detector_score": average(supports) + min(float(_HOUGH_CONFIG["areaScoreCap"]), area),
                    "detector_diagnostics": {
                        "edge_point_count": len(points),
                        "hough_peak_count": len(peaks),
                        "segment_count": len(segments),
                        "family_angle_degrees": round(math.degrees(family_angle), 2),
                        "line_support": [round(value, 3) for value in supports],
                        "gradient_threshold": round(gradient.threshold, 4),
                        "gradient_scales": list(gradient.scales),
                    },
                }
            )

    candidates.sort(key=lambda item: item["detector_score"], reverse=True)
    selected = []
    for candidate in candidates:
        if any(
            normalized_corner_distance(candidate["quad"], kept["quad"], width, height)
            < float(DETECTION_CONFIG["deduplication"]["cornerDistanceThreshold"])
            for kept in selected
        ):
            continue
        selected.append(candidate)
        if len(selected) >= int(_HOUGH_CONFIG["outputCandidateLimit"]):
            break
    return selected


def _edge_points(gradient: GradientMap) -> np.ndarray:
    ys, xs = np.nonzero(gradient.magnitude >= gradient.threshold)
    if len(xs) > int(_HOUGH_CONFIG["maximumEdgePointsBeforeStride"]):
        keep = np.arange(len(xs)) % int(_HOUGH_CONFIG["strideWhenDense"]) == 0
        xs = xs[keep]
        ys = ys[keep]
    magnitude = gradient.magnitude[ys, xs]
    orientation = np.mod(gradient.orientation[ys, xs], math.pi)
    return np.column_stack((xs, ys, magnitude, orientation))


def _vote(points: np.ndarray, width: int, height: int) -> list[tuple[int, int, float]]:
    diagonal = math.hypot(width, height)
    angle_bins = int(_HOUGH_CONFIG["angleBins"])
    rho_bins = math.ceil(diagonal * 2 / RHO_STEP) + 1
    accumulator = np.zeros((angle_bins, rho_bins), dtype=np.float64)
    center_indices = np.rint(points[:, 3] / ANGLE_STEP).astype(np.int32) % angle_bins
    weights = float(_HOUGH_CONFIG["voteWeightBase"]) + np.minimum(
        float(_HOUGH_CONFIG["voteMagnitudeCap"]), points[:, 2]
    )
    vote_radius = int(_HOUGH_CONFIG["voteAngleRadius"])
    for offset in range(-vote_radius, vote_radius + 1):
        angle_indices = (center_indices + offset) % angle_bins
        angles = angle_indices * ANGLE_STEP
        rho_indices = np.rint(
            (points[:, 0] * np.cos(angles) + points[:, 1] * np.sin(angles) + diagonal) / RHO_STEP
        ).astype(np.int32)
        np.add.at(accumulator, (angle_indices, rho_indices), weights)

    minimum_votes = max(float(_HOUGH_CONFIG["minimumVotes"]), min(width, height) * float(_HOUGH_CONFIG["minimumVotesRatio"]))
    angle_indices, rho_indices = np.nonzero(accumulator >= minimum_votes)
    candidates = sorted(
        (
            (int(angle_index), int(rho_index), float(accumulator[angle_index, rho_index]))
            for angle_index, rho_index in zip(angle_indices, rho_indices)
        ),
        key=lambda item: item[2],
        reverse=True,
    )
    peaks: list[tuple[int, int, float]] = []
    for candidate in candidates:
        if all(
            _circular_bin_distance(candidate[0], peak[0], angle_bins)
            > int(_HOUGH_CONFIG["peakAngleDistance"])
            or abs(candidate[1] - peak[1]) > int(_HOUGH_CONFIG["peakRhoDistance"])
            for peak in peaks
        ):
            peaks.append(candidate)
        if len(peaks) >= int(_HOUGH_CONFIG["maximumPeaks"]):
            break
    return peaks


def _extract_segments(
    peaks: list[tuple[int, int, float]],
    points: np.ndarray,
    width: int,
    height: int,
) -> list[Segment]:
    diagonal = math.hypot(width, height)
    candidates = []
    for angle_index, rho_index, _ in peaks:
        normal_angle = angle_index * ANGLE_STEP
        normal = np.array([math.cos(normal_angle), math.sin(normal_angle)])
        direction = np.array([-normal[1], normal[0]])
        rho = rho_index * RHO_STEP - diagonal
        distances = np.abs(points[:, 0] * normal[0] + points[:, 1] * normal[1] - rho)
        alignment = np.minimum(
            np.abs(points[:, 3] - normal_angle),
            math.pi - np.abs(points[:, 3] - normal_angle),
        )
        support_points = points[
            (distances <= float(_HOUGH_CONFIG["lineDistance"]))
            & (alignment <= math.radians(float(_HOUGH_CONFIG["orientationToleranceDegrees"])))
        ]
        if len(support_points) < int(_HOUGH_CONFIG["minimumEdgePoints"]):
            continue
        projections = np.sort(support_points[:, 0] * direction[0] + support_points[:, 1] * direction[1])
        split_indices = np.flatnonzero(
            np.diff(projections)
            > max(float(_HOUGH_CONFIG["maximumGapFloor"]), diagonal * float(_HOUGH_CONFIG["maximumGapRatio"]))
        ) + 1
        for group in np.split(projections, split_indices):
            if len(group) < 2:
                continue
            length = float(group[-1] - group[0])
            if length < diagonal * float(_HOUGH_CONFIG["minimumSegmentLengthRatio"]):
                continue
            candidates.append(
                Segment(
                    np.array([normal[0], normal[1], -rho]),
                    (normal_angle + math.pi / 2) % math.pi,
                    normal_angle,
                    rho,
                    length,
                    min(1.0, len(group) / max(1.0, length / 2)),
                )
            )

    candidates.sort(key=lambda segment: segment.length * segment.support, reverse=True)
    selected = []
    for segment in candidates:
        if all(
            _orientation_difference(segment.normal_angle, kept.normal_angle)
            > math.radians(float(_HOUGH_CONFIG["lineDeduplicationAngleDegrees"]))
            or abs(segment.rho - kept.rho) > float(_HOUGH_CONFIG["lineDeduplicationRho"])
            for kept in selected
        ):
            selected.append(segment)
        if len(selected) >= int(_HOUGH_CONFIG["maximumSegments"]):
            break
    return selected


def _cluster_directions(segments: list[Segment]) -> tuple[tuple[float, list[Segment]], tuple[float, list[Segment]]] | None:
    if len(segments) < 4:
        return None
    vectors = np.asarray([(math.cos(2 * segment.angle), math.sin(2 * segment.angle)) for segment in segments])
    first_center = vectors[0]
    second_center = vectors[np.argmax(np.sum((vectors - first_center) ** 2, axis=1))]
    centers = np.asarray([first_center, second_center])
    assignments = np.zeros(len(segments), dtype=np.int32)
    for _ in range(8):
        distances = np.sum((vectors[:, None, :] - centers[None, :, :]) ** 2, axis=2)
        assignments = np.argmin(distances, axis=1)
        for family in (0, 1):
            indices = np.flatnonzero(assignments == family)
            if not len(indices):
                continue
            weights = np.asarray(
                [
                    segments[index].length
                    * max(float(_HOUGH_CONFIG["supportWeightFloor"]), segments[index].support)
                    for index in indices
                ]
            )
            centers[family] = np.average(vectors[indices], axis=0, weights=weights)
    grouped = [[segment for index, segment in enumerate(segments) if assignments[index] == family] for family in (0, 1)]
    if any(len(group) < 2 for group in grouped):
        return None
    angles = [math.atan2(center[1], center[0]) * 0.5 % math.pi for center in centers]
    return (angles[0], grouped[0]), (angles[1], grouped[1])


def _line_pairs(segments: list[Segment], width: int, height: int) -> list[tuple[Segment, Segment]]:
    center = np.array([width / 2, height / 2])
    positioned = [
        (
            segment,
            float(segment.line[0] * center[0] + segment.line[1] * center[1] + segment.line[2]),
        )
        for segment in segments
    ]
    pairs = []
    for first_index, first in enumerate(positioned):
        for second in positioned[first_index + 1 :]:
            separation = abs(first[1] - second[1])
            if separation < min(width, height) * float(_HOUGH_CONFIG["pairSeparationRatio"]):
                continue
            straddles = first[1] * second[1] <= 0
            score = (
                first[0].support
                + second[0].support
                + (first[0].length + second[0].length) / math.hypot(width, height)
                + (0.45 if straddles else 0)
            )
            pairs.append(((first[0], second[0]), score))
    pairs.sort(key=lambda item: item[1], reverse=True)
    return [pair for pair, _ in pairs[: int(_HOUGH_CONFIG["maximumPairs"])] ]


def _order_quad(points: np.ndarray) -> np.ndarray:
    sums = points.sum(axis=1)
    differences = points[:, 0] - points[:, 1]
    return np.asarray(
        [points[np.argmin(sums)], points[np.argmax(differences)], points[np.argmax(sums)], points[np.argmin(differences)]]
    )


def _orientation_difference(first: float, second: float) -> float:
    difference = abs(first % math.pi - second % math.pi)
    return min(difference, math.pi - difference)


def _circular_bin_distance(first: int, second: int, count: int) -> int:
    difference = abs(first - second)
    return min(difference, count - difference)

File: site/app/detection/hough-lines.ts
import {
  geometryIsValid,
  lineIntersection,
  orderQuad,
  polygonArea,
  type Line,
} from "./geometry.ts";
import { DETECTION_CONFIG } from "./config.ts";
import type {
  CandidateDetector,
  CandidateFeatures,
  ImageFeatures,
  Point,
  Quad,
  QuadCandidate,
} from "./types.ts";

const HOUGH_CONFIG = DETECTION_CONFIG.houghLines;

type EdgePoint = {
  x: number;
  y: number;
  magnitude: number;
  orientation: number;
};

type LineSegment = {
  line: Line;
  start: Point;
  end: Point;
  angle: number;
  normalAngle: number;
  rho: number;
  length: number;
  support: number;
};

type HoughPeak = {
  angleIndex: number;
  rhoIndex: number;
  votes: number;
};

const EMPTY_FEATURES: CandidateFeatures = {
  edgeStrength: 0,
  edgeSupport: 0,
  edgeContinuity: 0,
  gradientAlignment: 0,
  insideOutsideDifference: 0,
  regionConsistency: 0,
  normalizedArea: 0,
  geometryValidity: 0,
  aspectPrior: 0,
  batchConsistency: 0,
};

const ANGLE_STEP = Math.PI / HOUGH_CONFIG.angleBins;
const RHO_STEP = HOUGH_CONFIG.rhoStep;

export const houghLineDetector: CandidateDetector = {
  name: "hough-lines",
  detect(features: ImageFeatures): QuadCandidate[] {
    const edgePoints = collectEdgePoints(features);
    if (edgePoints.length < HOUGH_CONFIG.minimumEdgePoints) return [];
    const peaks = voteForLines(edgePoints, features.width, features.height);
    const segments = extractSegments(peaks, edgePoints, features.width, features.height);
    const families = clusterDirections(segments);
    if (!families) return [];
    const [firstFamily, secondFamily] = families;
    const familyAngle = directionDifference(firstFamily.angle, secondFamily.angle);
    if (
      familyAngle < HOUGH_CONFIG.minimumFamilyAngleDegrees * Math.PI / 180 ||
      familyAngle > HOUGH_CONFIG.maximumFamilyAngleDegrees * Math.PI / 180
    ) return [];

    const firstPairs = generateLinePairs(firstFamily.segments, features.width, features.height);
    const secondPairs = generateLinePairs(secondFamily.segments, features.width, features.height);
    const candidates: Array<{ quad: Quad; detectorScore: number; lineSupport: number[] }> = [];
    for (const firstPair of firstPairs) {
      for (const secondPair of secondPairs) {
        const quad = orderQuad([
          lineIntersection(firstPair[0].line, secondPair[0].line),
          lineIntersection(firstPair[0].line, secondPair[1].line),
          lineIntersection(firstPair[1].line, secondPair[1].line),
          lineIntersection(firstPair[1].line, secondPair[0].line),
        ]);
        if (!geometryIsValid(quad, features.width, features.height)) continue;
        const area = polygonArea(quad) / (features.width * features.height);
        const lineSupport = [...firstPair, ...secondPair].map((segment) => segment.support);
        const detectorScore = average(lineSupport) + Math.min(HOUGH_CONFIG.areaScoreCap, area);
        candidates.push({ quad, detectorScore, lineSupport });
      }
    }

    return candidates
      .sort((first, second) => second.detectorScore - first.detectorScore)
      .filter((candidate, index, all) =>
        all.findIndex((other) =>
          quadDistance(candidate.quad, other.quad, features.width, features.height) <
            DETECTION_CONFIG.deduplication.cornerDistanceThreshold
        ) ===
          index
      )
      .slice(0, HOUGH_CONFIG.outputCandidateLimit)
      .map((candidate) => ({
        quad: candidate.quad,
        method: "hough-lines",
        polarity: [],
        features: { ...EMPTY_FEATURES },
        rawScore: candidate.detectorScore,
        warnings: [],
        diagnostics: {
          edgePointCount: edgePoints.length,
          houghPeakCount: peaks.length,
          segmentCount: segments.length,
          familyAngleDegrees: round(familyAngle * 180 / Math.PI, 2),
          lineSupport: candidate.lineSupport.map((value) => round(value, 3)),
          gradientThreshold: round(features.gradient.threshold, 4),
          gradientScales: features.gradient.scales,
        },
      }));
  },
};

function collectEdgePoints(features: ImageFeatures): EdgePoint[] {
  const { magnitude, orientation, threshold } = features.gradient;
  const rawCount = magnitude.reduce((count, value) => count + (value >= threshold ? 1 : 0), 0);
  const stride = rawCount > HOUGH_CONFIG.maximumEdgePointsBeforeStride
    ? HOUGH_CONFIG.strideWhenDense
    : 1;
  const points: EdgePoint[] = [];
  for (let y = 1; y < features.height - 1; y += stride) {
    for (let x = 1; x < features.width - 1; x += stride) {
      const index = y * features.width + x;
      if (magnitude[index] < threshold) continue;
      points.push({ x, y, magnitude: magnitude[index], orientation: normalizeAngle(orientation[index]) });
    }
  }
  return points;
}

function voteForLines(edgePoints: EdgePoint[], width: number, height: number): HoughPeak[] {
  const diagonal = Math.hypot(width, height);
  const angleBins = HOUGH_CONFIG.angleBins;
  const rhoBins = Math.ceil(diagonal * 2 / RHO_STEP) + 1;
  const accumulator = new Float64Array(angleBins * rhoBins);
  for (const point of edgePoints) {
    const centerIndex = Math.round(point.orientation / ANGLE_STEP) % angleBins;
    for (let offset = -HOUGH_CONFIG.voteAngleRadius; offset <= HOUGH_CONFIG.voteAngleRadius; offset += 1) {
      const angleIndex = modulo(centerIndex + offset, angleBins);
      const angle = angleIndex * ANGLE_STEP;
      const rho = point.x * Math.cos(angle) + point.y * Math.sin(angle);
      const rhoIndex = Math.round((rho + diagonal) / RHO_STEP);
      accumulator[angleIndex * rhoBins + rhoIndex] +=
        HOUGH_CONFIG.voteWeightBase + Math.min(HOUGH_CONFIG.voteMagnitudeCap, point.magnitude);
    }
  }

  const peakCandidates: HoughPeak[] = [];
  for (let angleIndex = 0; angleIndex < angleBins; angleIndex += 1) {
    for (let rhoIndex = 0; rhoIndex < rhoBins; rhoIndex += 1) {
      const votes = accumulator[angleIndex * rhoBins + rhoIndex];
      if (votes >= Math.max(HOUGH_CONFIG.minimumVotes, Math.min(width, height) * HOUGH_CONFIG.minimumVotesRatio)) {
        peakCandidates.push({ angleIndex, rhoIndex, votes });
      }
    }
  }
  peakCandidates.sort((first, second) => second.votes - first.votes);
  const peaks: HoughPeak[] = [];
  for (const candidate of peakCandidates) {
    if (peaks.every((peak) =>
      circularBinDistance(candidate.angleIndex, peak.angleIndex, angleBins) > HOUGH_CONFIG.peakAngleDistance ||
      Math.abs(candidate.rhoIndex - peak.rhoIndex) > HOUGH_CONFIG.peakRhoDistance
    )) {
      peaks.push(candidate);
    }
    if (peaks.length >= HOUGH_CONFIG.maximumPeaks) break;
  }
  return peaks;
}

function extractSegments(
  peaks: HoughPeak[],
  edgePoints: EdgePoint[],
  width: number,
  height: number,
): LineSegment[] {
  const diagonal = Math.hypot(width, height);
  const candidates: LineSegment[] = [];
  for (const peak of peaks) {
    const normalAngle = peak.angleIndex * ANGLE_STEP;
    const normalX = Math.cos(normalAngle);
    const normalY = Math.sin(normalAngle);
    const directionX = -normalY;
    const directionY = normalX;
    const rho = peak.rhoIndex * RHO_STEP - diagonal;
    const aligned = edgePoints
      .filter((point) =>
        Math.abs(point.x * normalX + point.y * normalY - rho) <= HOUGH_CONFIG.lineDistance &&
        orientationDifference(point.orientation, normalAngle) <= HOUGH_CONFIG.orientationToleranceDegrees * Math.PI / 180
      )
      .map((point) => ({ point, projection: point.x * directionX + point.y * directionY }))
      .sort((first, second) => first.projection - second.projection);
    if (aligned.length < HOUGH_CONFIG.minimumEdgePoints) continue;

    const groups: typeof aligned[] = [];
    let group: typeof aligned = [aligned[0]];
    const maximumGap = Math.max(HOUGH_CONFIG.maximumGapFloor, diagonal * HOUGH_CONFIG.maximumGapRatio);
    for (let index = 1; index < aligned.length; index += 1) {
      if (aligned[index].projection - aligned[index - 1].projection > maximumGap) {
        groups.push(group);
        group = [];
      }
      group.push(aligned[index]);
    }
    groups.push(group);

    for (const support of groups) {
      const first = support[0].projection;
      const last = support[support.length - 1].projection;
      const length = last - first;
      if (length < diagonal * HOUGH_CONFIG.minimumSegmentLengthRatio) continue;
      candidates.push({
        line: { a: normalX, b: normalY, c: -rho },
        start: [normalX * rho + directionX * first, normalY * rho + directionY * first],
        end: [normalX * rho + directionX * last, normalY * rho + directionY * last],
        angle: normalizeAngle(normalAngle + Math.PI / 2),
        normalAngle,
        rho,
        length,
        support: Math.min(1, support.length / Math.max(1, length / 2)),
      });
    }
  }
  candidates.sort((first, second) => second.length * second.support - first.length * first.support);
  const selected: LineSegment[] = [];
  for (const segment of candidates) {
    if (selected.every((kept) =>
      orientationDifference(segment.normalAngle, kept.normalAngle) >
        HOUGH_CONFIG.lineDeduplicationAngleDegrees * Math.PI / 180 ||
      Math.abs(segment.rho - kept.rho) > HOUGH_CONFIG.lineDeduplicationRho
    )) {
      selected.push(segment);
    }
    if (selected.length >= HOUGH_CONFIG.maximumSegments) break;
  }
  return selected;
}

function clusterDirections(segments: LineSegment[]): [
  { angle: number; segments: LineSegment[] },
  { angle: number; segments: LineSegment[] },
] | null {
  if (segments.length < 4) return null;
  const vectors = segments.map((segment) => [Math.cos(2 * segment.angle), Math.sin(2 * segment.angle)] as Point);
  let centers: [Point, Point] = [
    [...vectors[0]] as Point,
    [...vectors[vectors.reduce((best, vector, index) =>
      squaredDistance(vector, vectors[0]) > squaredDistance(vectors[best], vectors[0]) ? index : best
    , 0)]] as Point,
  ];
  let assignments = new Array<number>(segments.length).fill(0);
  for (let iteration = 0; iteration < 8; iteration += 1) {
    assignments = vectors.map((vector) =>
      squaredDistance(vector, centers[0]) <= squaredDistance(vector, centers[1]) ? 0 : 1
    );
    centers = [0, 1].map((family) => {
      let x = 0;
      let y = 0;
      let weight = 0;
      segments.forEach((segment, index) => {
        if (assignments[index] !== family) return;
        const itemWeight = segment.length * Math.max(HOUGH_CONFIG.supportWeightFloor, segment.support);
        x += vectors[index][0] * itemWeight;
        y += vectors[index][1] * itemWeight;
        weight += itemWeight;
      });
      return weight ? [x / weight, y / weight] as Point : centers[family];
    }) as [Point, Point];
  }
  const grouped = [0, 1].map((family) => segments.filter((_, index) => assignments[index] === family));
  if (grouped.some((group) => group.length < 2)) return null;
  return [
    { angle: normalizeAngle(0.5 * Math.atan2(centers[0][1], centers[0][0])), segments: grouped[0] },
    { angle: normalizeAngle(0.5 * Math.atan2(centers[1][1], centers[1][0])), segments: grouped[1] },
  ];
}

function generateLinePairs(segments: LineSegment[], width: number, height: number): Array<[LineSegment, LineSegment]> {
  const center: Point = [width / 2, height / 2];
  const positioned = segments.map((segment) => ({
    segment,
    position: segment.line.a * center[0] + segment.line.b * center[1] + segment.line.c,
  }));
  const pairs: Array<{ pair: [LineSegment, LineSegment]; score: number }> = [];
  for (let first = 0; first < positioned.length; first += 1) {
    for (let second = first + 1; second < positioned.length; second += 1) {
      const separation = Math.abs(positioned[first].position - positioned[second].position);
      if (separation < Math.min(width, height) * HOUGH_CONFIG.pairSeparationRatio) continue;
      const straddlesCenter = positioned[first].position * positioned[second].position <= 0;
      const support = positioned[first].segment.support + positioned[second].segment.support;
      const length = positioned[first].segment.length + positioned[second].segment.length;
      pairs.push({
        pair: [positioned[first].segment, positioned[second].segment],
        score: support + length / Math.hypot(width, height) + (straddlesCenter ? 0.45 : 0),
      });
    }
  }
  return pairs
    .sort((first, second) => second.score - first.score)
    .slice(0, HOUGH_CONFIG.maximumPairs)
    .map((item) => item.pair);
}

function normalizeAngle(angle: number): number {
  return modulo(angle, Math.PI);
}

function orientationDifference(first: number, second: number): number {
  const difference = Math.abs(normalizeAngle(first) - normalizeAngle(second));
  return Math.min(difference, Math.PI - difference);
}

function directionDifference(first: number, second: number): number {
  return orientationDifference(first, second);
}

function circularBinDistance(first: number, second: number, count: number): number {
  const difference = Math.abs(first - second);
  return Math.min(difference, count - difference);
}

function modulo(value: number, modulus: number): number {
  return ((value % modulus) + modulus) % modulus;
}

function squaredDistance(first: Point, second: Point): number {
  return (first[0] - second[0]) ** 2 + (first[1] - second[1]) ** 2;
}

function quadDistance(first: Quad, second: Quad, width: number, height: number): number {
  const diagonal = Math.hypot(width, height);
  return average(first.map((point, index) => Math.hypot(point[0] - second[index][0], point[1] - second[index][1]))) /
    diagonal;
}

function average(values: ArrayLike<number>): number {
  if (!values.length) return 0;
  let total = 0;
  for (let index = 0; index < values.length; index += 1) total += values[index];
  return total / values.length;
}

function round(value: number, digits: number): number {
  const scale = 10 ** digits;
  return Math.round(value * scale) / scale;
}
